### PR TITLE
[JIT] Shape inference improvement

### PR DESCRIPTION
**Pull request by jjsjann123**
Support `aten::div` in `PropagateCompleteShapeOnNode`.

complete shape propagation on `aten::div` is disabled, because shape inference
relies on running node to propagate shape. For `aten::div` we run into
deviding-by-zero problem.

However, shape propagation for pointwise operatoins should be identical. We
would be able to swap the operation for `aten::div` with `aten::mul`.

### CODE REVIEW

**Comment by jjsjann123** [comment]
This is to replace the old PR #35031 where the fix was based on a wrong commit.

**Comment by yf225** [comment]
@suo Curious would anyone on the JIT team be interested in reviewing this PR?

**Comment by suo** [comment]
cc @eellison, do you want to take a look?

File: torch/csrc/jit/passes/shape_analysis.cpp [hunk 1]
```diff
@@ -1823,11 +1826,18 @@ class ShapePropagator {
...
+      // These nodes handle tensors of different shapes internally, so there's
+      // no need to insert explicit expand nodes.
       return PropagateShapeOnNodeByRunningIt(node);
+    } else if (node->matches("aten::div(Tensor self, Tensor other) -> Tensor")) {
+      // "div" handle tensors of different shapes internally, so there's no need
+      // to insert explicit expand nodes.
+      // Note that this function could be merged to the one above , but "div" is
+      // not always safe to run by itself due to integer divide-by-zero.
+      // We fake the execution by running "mul" operation instead.
+      auto op = getOperatorForLiteral(
```
Comment (eellison) [inline torch/csrc/jit/passes/shape_analysis.cpp:1838]: Are we sure that `mul` and `div` have the same dtype promotion rules?

Comment (jjsjann123) [reply torch/csrc/jit/passes/shape_analysis.cpp:1838]:
```
import torch

def t_promo(l_dtype, r_dtype):
    x = torch.ones(5, 5, dtype=l_dtype, device="cuda")
    y = torch.ones(5, 5, dtype=r_dtype, device="cuda")
    out_div = x / y
    out_mul = x * y
    print(out_div.dtype, ", ", out_mul.dtype)
    assert(out_div.dtype == out_mul.dtype)

# no bool tested
dtypes = [torch.float32, torch.float64, torch.float16, torch.uint8, torch.int8, torch.int16, torch.int32, torch.int64]

for l in dtypes:
  for r in dtypes:
    t_promo(l, r)

```

Maybe someone who knows details in type promotion would be able to answer this better.
Empirically they seem to converge and I can't think of a good reason for them to differ.

**Review by eellison — APPROVED** [review]
LGTM

**Comment by eellison** [comment]
@jjsjann123 mind rebasing & i'll land ? sorry review took a while.

**Comment by jjsjann123** [comment]
Updated.

**Comment by eellison** [comment]
clang format lint is breaking on shape analysis :/ kind of annoying I know

**Comment by jjsjann123** [comment]
Oops, I did look at the clang-format but might have missed that line with the error. :stuck_out_tongue:

**Comment by jjsjann123** [comment]
Didn't see any related failure. Would be strange that clang-format breaks code...

**Comment by eellison** [comment]
yea i think the failures are upstream. i'm landing this

**Pull request closed without merging**